### PR TITLE
Heretic: fix Tutti Frutti effect

### PR DESCRIPTION
**Pull request by JNechaevsky**
PR system seems to working now (I was having some odd problems this morning), so here it is.

Map for testing: http://jnechaevsky.users.sourceforge.net/files/tutti-heretic.wad
Just take a try before/after.

This also fixes "pixel disco" effect, remember?

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
@JNechaevsky I have already left quite some review comments.

**Comment by JNechaevsky** [comment]
But where exactly, could you please point out? Either I'm blind or I lost my way in three pines.

**Comment by fabiangreffrath** [comment]
Just have a look at this page:
https://github.com/fabiangreffrath/crispy-doom/pull/277/files

**Comment by JNechaevsky** [comment]
You have to forgive me, but I don't see any your comments, there are only my changes in the code.
I'm not kidding. Could it be another f...-up of GitHub?

File: src/heretic/r_draw.c [hunk 1]
```diff
@@ -50,17 +51,19 @@ int dc_yl;
...
 {
     int count;
-    byte *dest;
-    fixed_t frac, fracstep;
+    register byte *dest;        // killough
+    register fixed_t frac;      // killough
+    fixed_t fracstep;
 
-    count = dc_yh - dc_yl;
+    count = dc_yh - dc_yl + 1;
```
[the remark this answers is not in the document]
Comment (JNechaevsky) [reply src/heretic/r_draw.c]: Right from initial fixing commit: https://github.com/fabiangreffrath/crispy-doom/commit/eaf48d3e656197084a476295f56f3530f9775626#diff-7c44312d8e90e8ade465c6af58116e47R113.

File: src/heretic/r_draw.c [hunk 2]
```diff
@@ -32,6 +32,7 @@ byte *viewimage;
 int viewwidth, scaledviewwidth, viewheight, viewwindowx, viewwindowy;
 byte *ylookup[MAXHEIGHT];
 int columnofs[MAXWIDTH];
+int linesize = SCREENWIDTH;
```
[the remark this answers is not in the document]
Comment (JNechaevsky) [reply src/heretic/r_draw.c]: It should be just `int linesize;` then, right?

File: src/heretic/r_draw.c [hunk 3]
```diff
@@ -69,18 +72,67 @@ void R_DrawColumn(void)
...
+    // This is as fast as it gets.       (Yeah, right!!! -- killough)
+    //
+    // killough 2/1/98: more performance tuning
+
     {
-        *dest = dc_colormap[dc_source[(frac >> FRACBITS) & 127]];
-        dest += SCREENWIDTH;
-        frac += fracstep;
+        register const byte *source = dc_source;
+        register const lighttable_t *colormap = dc_colormap;
```
[the remark this answers is not in the document]
Comment (JNechaevsky) [reply src/heretic/r_draw.c]: So?
```
    source = dc_source;
    colormap = dc_colormap;	
```

File: src/heretic/r_draw.c [hunk 4]
```diff
@@ -69,18 +72,67 @@ void R_DrawColumn(void)
...
+                frac += fracstep;
+            }
+            if (count & 1)
+                *dest = colormap[source[(frac>>FRACBITS) & heightmask]];
+        }
     }
-    while (count--);
 }
 
 void R_DrawColumnLow(void)
```
[the remark this answers is not in the document]
Comment (JNechaevsky) [reply src/heretic/r_draw.c]: :O But Heretic does not have low detail mode at all, this `R_DrawColumnLow` seems to be a initial leftover from Doom code. I haven't added. What should we do?

File: src/heretic/r_draw.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (fabiangreffrath) [reply src/heretic/r_draw.c]: Nope, please use `SCREENWIDTH` where there's currently `linesize` in the code.

File: src/heretic/r_draw.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabiangreffrath) [reply src/heretic/r_draw.c]: Nope, just use `dc_source` where there is currently `source` used in the code, same for `dc_colormap` and `colormap`. It makes no sense to re-define variables all over again to have them marked as `register`.

File: src/heretic/r_draw.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabiangreffrath) [reply src/heretic/r_draw.c]: Oh, I forgot, then we can leave it as it is. ;)

**Pull request merged**